File: otb/src/main/java/com/orange/essentials/otb/OtbActivity.java
/*
 * Copyright 2016 Orange
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* Orange Trust Badge library
 *
 * Module name: com.orange.essentials:otb
 * Version:     1.0
 * Created:     2016-03-15 by Aurore Penault, Vincent Boesch, and Giovanni Battista Accetta
 */
package com.orange.essentials.otb;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;

import com.orange.essentials.otb.event.EventType;
import com.orange.essentials.otb.manager.BadgeListener;
import com.orange.essentials.otb.manager.TrustBadgeManager;
import com.orange.essentials.otb.model.Term;
import com.orange.essentials.otb.model.TrustBadgeElement;
import com.orange.essentials.otb.model.type.GroupType;
import com.orange.essentials.otb.ui.OtbContainerFragment;
import com.orange.essentials.otb.ui.OtbDataFragment;
import com.orange.essentials.otb.ui.OtbTermsFragment;
import com.orange.essentials.otb.ui.OtbUsageFragment;

import java.io.Serializable;
import java.util.List;


/**
 * OtbActivity
 * Main activity for the lib
 */

public class OtbActivity extends AppCompatActivity implements OtbContainerFragment.OtbFragmentListener, BadgeListener {

    private static final String TAG = "OtbActivity";
    private static final String BADGES_KEY = "BadgesKey";
    private static final String TERMS_KEY = "TermsKey";

    private static boolean isMasterDetail;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.otb_activity);

        /** Manage toolbar */
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setTitle(R.string.otb_app_name);
        }

        /** Calling the fragment */
        if (savedInstanceState == null) {
            isMasterDetail = useMasterDetail();
            initFragments();
        } else {
            Log.v(TAG, "savedInstanceState != null");
            Log.v(TAG, "useMasterDetail: " + useMasterDetail());
            if (useMasterDetail() != isMasterDetail) {
                Log.v(TAG, "popBackstack");
                //restore from scratch
                FragmentManager fm = getSupportFragmentManager();
                for (int i = 0; i < fm.getBackStackEntryCount(); ++i) {
                    fm.popBackStack();
                }
                initFragments();
                isMasterDetail = useMasterDetail();
            }
        }

        TrustBadgeManager.INSTANCE.addBadgeListener(this);
    }

    private void initFragments() {
        Fragment otbContainerFrag = getSupportFragmentManager().findFragmentByTag(OtbContainerFragment.FRAG_TAG);
        if (otbContainerFrag == null) {
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.lightfragment_container, OtbContainerFragment.newInstance(), OtbContainerFragment.FRAG_TAG)
                    .commit();
        } else {
            getSupportFragmentManager().beginTransaction()
                    .attach(otbContainerFrag)
                    .commit();
        }

        // Landscape mode
        if (useMasterDetail()) {
            Log.d(TAG, "Landscape mode - add data fragment");
            OtbDataFragment frag = new OtbDataFragment();
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.lightfragment_detail, frag, OtbDataFragment.FRAG_TAG)
                    .commit();

        }
    }

    // region OtbContainerFragment.ContainerFragmentListener

    @Override
    public void onDataClick() {
        Log.d(TAG, "onDataClick");
        displayDetail(new OtbDataFragment(), OtbDataFragment.FRAG_TAG);
    }

    @Override
    public void onUsageClick() {
        Log.d(TAG, "onUsageClick");
        displayDetail(new OtbUsageFragment(), OtbUsageFragment.FRAG_TAG);
    }

    @Override
    public void onTermsClick() {
        Log.d(TAG, "onTermsClick");
        displayDetail(new OtbTermsFragment(), OtbTermsFragment.FRAG_TAG);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            // app icon in action bar clicked; goto parent activity.
            FragmentManager fm = getSupportFragmentManager();
            if (fm.getBackStackEntryCount() > 0) {
                fm.popBackStack();
            } else {
                finish();
            }
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onDestroy() {
        TrustBadgeManager.INSTANCE.getEventTagger().tag(EventType.TRUSTBADGE_LEAVE);
        super.onDestroy();
    }

    // endregion


    // region Private methods
    private boolean useMasterDetail() {
        return findViewById(R.id.lightfragment_detail) != null;
    }

    private void displayDetail(Fragment fragment, String tag) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        if (useMasterDetail()) {
            transaction.replace(R.id.lightfragment_detail, fragment, tag);
        } else {
            transaction.replace(R.id.lightfragment_container, fragment, tag);
            transaction.addToBackStack(null);
        }
        transaction.commit();
    }

    @Override
    public void onBadgeChange(TrustBadgeElement trustBadgeElement, boolean value, AppCompatActivity callingActivity) {
        Log.d(TAG, "onChange trustBadgeElement=" + trustBadgeElement + " value=" + value);
        if( null != trustBadgeElement ) {
            if (GroupType.IMPROVEMENT_PROGRAM.equals(trustBadgeElement.getGroupType())) {
                TrustBadgeManager.INSTANCE.setUsingImprovementProgram(value);
                Fragment frag;
                if (useMasterDetail()) {
                    frag = getSupportFragmentManager().findFragmentById(R.id.lightfragment_detail);
                } else {
                    frag = getSupportFragmentManager().findFragmentById(R.id.lightfragment_container);
                }
                if (frag instanceof OtbDataFragment) {
                    ((OtbDataFragment) frag).refreshPermission();
                }
            }
        }
        //No action defined for other badges
    }
    // endregion


    @Override
    protected void onSaveInstanceState(Bundle outState) {
        Log.d(TAG, "Saving Factory");
        super.onSaveInstanceState(outState);
        outState.putSerializable(BADGES_KEY, (Serializable) TrustBadgeManager.INSTANCE.getTrustBadgeElements());
        outState.putSerializable(TERMS_KEY, (Serializable) TrustBadgeManager.INSTANCE.getTerms());
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        Log.d(TAG, "Restoring factory");
        super.onRestoreInstanceState(savedInstanceState);
        restoreFactory(savedInstanceState);
    }

    private void restoreFactory(Bundle savedInstanceState) {
        if (null != savedInstanceState && null != savedInstanceState.getSerializable(BADGES_KEY)) {
            Log.d(TAG, "Restoring factory from instanceState");
            List<TrustBadgeElement> badges = (List<TrustBadgeElement>) savedInstanceState.getSerializable(BADGES_KEY);
            List<Term> terms = (List<Term>) savedInstanceState.getSerializable(TERMS_KEY);
            TrustBadgeManager.INSTANCE.initialize(getApplicationContext(), badges, terms);
        }
    }
}

File: otb/src/main/java/com/orange/essentials/otb/OtbActivity.kt
/*
 * Copyright 2016 Orange
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* Orange Trust Badge library
 *
 * Module name: com.orange.essentials:otb
 * Version:     1.0
 * Created:     2016-03-15 by Aurore Penault, Vincent Boesch, and Giovanni Battista Accetta
 */
package com.orange.essentials.otb

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.MenuItem
import android.view.View
import com.orange.essentials.otb.event.EventType
import com.orange.essentials.otb.manager.BadgeListener
import com.orange.essentials.otb.manager.TrustBadgeManager
import com.orange.essentials.otb.model.Term
import com.orange.essentials.otb.model.TrustBadgeElement
import com.orange.essentials.otb.model.type.GroupType
import com.orange.essentials.otb.ui.OtbAppDataFragment
import com.orange.essentials.otb.ui.OtbContainerFragment
import com.orange.essentials.otb.ui.OtbPermissionsFragment
import com.orange.essentials.otb.ui.OtbTermsFragment
import java.io.Serializable

/**
 * OtbActivity
 * Main activity for the lib
 */
open class OtbActivity : AppCompatActivity(), OtbContainerFragment.OtbFragmentListener, BadgeListener {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.otb_activity)
        /** Manage toolbar  */
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setTitle(R.string.otb_app_name)
        /** Calling the fragment  */
        if (savedInstanceState == null) {
            isMasterDetail = useMasterDetail()
            initFragments()
        } else {
            Log.v(TAG, "savedInstanceState != null")
            Log.v(TAG, "useMasterDetail: " + useMasterDetail())
            if (useMasterDetail() != isMasterDetail) {
                Log.v(TAG, "popBackstack")
                //restore from scratch
                val fm = supportFragmentManager
                for (i in 0 until fm.backStackEntryCount) {
                    fm.popBackStack()
                }
                initFragments()
                isMasterDetail = useMasterDetail()
            }
        }

        TrustBadgeManager.INSTANCE.addBadgeListener(this)
    }

    private fun initFragments() {
        val otbContainerFrag = supportFragmentManager.findFragmentByTag(OtbContainerFragment.FRAG_TAG)
        if (otbContainerFrag == null) {
            supportFragmentManager.beginTransaction()
                    .add(R.id.lightfragment_container, OtbContainerFragment.newInstance(), OtbContainerFragment.FRAG_TAG)
                    .commit()
        } else {
            supportFragmentManager.beginTransaction()
                    .attach(otbContainerFrag)
                    .commit()
        }
        // Landscape mode
        if (useMasterDetail()) {
            Log.d(TAG, "Landscape mode - add child fragment")
            val frag: Fragment?
            val tag: String?
            when {
                TrustBadgeManager.INSTANCE.hasPermissions() -> {
                    Log.d(TAG, "Landscape mode - add data fragment")
                    frag = OtbPermissionsFragment()
                    tag = OtbPermissionsFragment.FRAG_TAG
                }
                TrustBadgeManager.INSTANCE.hasAppData() -> {
                    Log.d(TAG, "Landscape mode - add usage fragment")
                    frag = OtbAppDataFragment()
                    tag = OtbAppDataFragment.FRAG_TAG
                }
                TrustBadgeManager.INSTANCE.hasTerms() -> {
                    Log.d(TAG, "Landscape mode - add terms fragment")
                    frag = OtbTermsFragment()
                    tag = OtbTermsFragment.FRAG_TAG
                }
                else -> {
                    Log.d(TAG, "Landscape mode - No item found, add data fragment by default")
                    frag = OtbPermissionsFragment()
                    tag = OtbPermissionsFragment.FRAG_TAG
                }
            }
            supportFragmentManager.beginTransaction()
                    .add(R.id.lightfragment_detail, frag, tag)
                    .commit()

        }
    }

    // region OtbContainerFragment.ContainerFragmentListener
    override fun onCardClick(index: Int) {
        Log.d(TAG, "onCardClick, index $index")
        when (index) {
            OtbContainerFragment.DATA_SELECTED -> {
                displayDetail(OtbPermissionsFragment(), OtbPermissionsFragment.FRAG_TAG)
                TrustBadgeManager.INSTANCE.eventTagger?.tag(EventType.TRUSTBADGE_GO_TO_PERMISSION)
            }
            OtbContainerFragment.USAGE_SELECTED -> {
                displayDetail(OtbAppDataFragment(), OtbAppDataFragment.FRAG_TAG)
                TrustBadgeManager.INSTANCE.eventTagger?.tag(EventType.TRUSTBADGE_GO_TO_USAGE)
            }
            OtbContainerFragment.TERM_SELECTED -> {
                displayDetail(OtbTermsFragment(), OtbTermsFragment.FRAG_TAG)
                TrustBadgeManager.INSTANCE.eventTagger?.tag(EventType.TRUSTBADGE_GO_TO_TERMS)
            }
            else -> displayDetail(TrustBadgeManager.INSTANCE.customDataFragments[index - 3].fragment, TrustBadgeManager.INSTANCE.customDataFragments[index - 3].fragment.tag.toString())
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when {
            (item.itemId == android.R.id.home) -> {
                val fm = supportFragmentManager
                if (fm.backStackEntryCount > 0) {
                    fm.popBackStack()
                } else {
                    finish()
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroy() {
        TrustBadgeManager.INSTANCE.eventTagger?.tag(EventType.TRUSTBADGE_LEAVE)
        super.onDestroy()
    }

    // endregion
    // region Private methods
    private fun useMasterDetail(): Boolean {
        return findViewById<View>(R.id.lightfragment_detail) != null
    }

    private fun displayDetail(fragment: Fragment, tag: String) {
        val transaction = supportFragmentManager.beginTransaction()
        if (useMasterDetail()) {
            transaction.replace(R.id.lightfragment_detail, fragment, tag)
        } else {
            transaction.replace(R.id.lightfragment_container, fragment, tag)
            transaction.addToBackStack(null)
        }
        transaction.commit()
    }

    override fun onBadgeChange(trustBadgeElement: TrustBadgeElement, value: Boolean, callingActivity: AppCompatActivity) {
        Log.d(TAG, "onChange trustBadgeElement=$trustBadgeElement value=$value")
        when (trustBadgeElement.groupType) {
            GroupType.IMPROVEMENT_PROGRAM -> {
                refreshAll(callingActivity)
            }
            GroupType.CUSTOM_DATA -> {
                refreshAll(callingActivity)
            }
            else -> {
            }
        }
    }

    private fun refreshAll(callingActivity: AppCompatActivity) {
        val frag = if (useMasterDetail()) {
            callingActivity.supportFragmentManager.findFragmentById(R.id.lightfragment_detail)
        } else {
            callingActivity.supportFragmentManager.findFragmentById(R.id.lightfragment_container)
        }
        if (frag is OtbPermissionsFragment) {
            frag.refreshPermissions()
        }
        if (frag is OtbAppDataFragment) {
            frag.refreshAppData()
        }
    }

    // endregion
    override fun onSaveInstanceState(outState: Bundle) {
        Log.d(TAG, "Saving Factory")
        super.onSaveInstanceState(outState)
        outState.putSerializable(BADGES_KEY, TrustBadgeManager.INSTANCE.trustBadgeElements as Serializable)
        outState.putSerializable(TERMS_KEY, TrustBadgeManager.INSTANCE.terms as Serializable)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        Log.d(TAG, "Restoring factory")
        restoreFactory(savedInstanceState)
        super.onRestoreInstanceState(savedInstanceState)
    }

    @Suppress("UNCHECKED_CAST")
    private fun restoreFactory(savedInstanceState: Bundle?) {
        savedInstanceState?.getSerializable(BADGES_KEY)?.let {
            Log.d(TAG, "Restoring factory from instanceState")
            val badges = savedInstanceState.getSerializable(BADGES_KEY) as MutableList<TrustBadgeElement>
            val terms = savedInstanceState.getSerializable(TERMS_KEY) as List<Term>
            TrustBadgeManager.INSTANCE.initialize(applicationContext, badges, terms)
        }
    }

    override fun onResume() {
        super.onResume()
        TrustBadgeManager.INSTANCE.eventTagger?.tag(EventType.TRUSTBADGE_ENTER)
    }

    companion object {
        private const val TAG = "OtbActivity"
        private const val BADGES_KEY = "BadgesKey"
        private const val TERMS_KEY = "TermsKey"
        var isMasterDetail: Boolean = false
    }
}
